Sources/X509/TinyArray.swift

```diff
@@ -0,0 +1,323 @@
+        }
+    }
+}
+extension TinyArray.Storage: Sendable where Element: Sendable {}
+
+extension TinyArray.Storage: RandomAccessCollection {
+    @inlinable
+    subscript(position: Int) -> Element {
+        get {
+            switch self {
+            case .one(let element):
+                guard position == 0 else {
+                    fatalError("index \(position) out of bounds")
+                }
+                return element
+            case .arbitrary(let elements):
+                return elements[position]
+            }
+        }
+    }
+    
+    @inlinable
+    var startIndex: Int {
+        0
+    }
+    
+    @inlinable
+    var endIndex: Int {
+        switch self {
+        case .one: return 1
+        case .arbitrary(let elements): return elements.endIndex
+        }
+    }
+}
+ 
+extension TinyArray.Storage {
+    @inlinable
+    init(_ elements: some Sequence<Element>) {
+        self = .arbitrary([])
+        self.append(contentsOf: elements)
+    }
+    
+    @inlinable
+    init(_ newElements: some Sequence<Result<Element, some Error>>) throws {
+        var iterator = newElements.makeIterator()
+        guard let firstElement = try iterator.next()?.get() else {
+            self = .arbitrary([])
+            return
+        }
+        guard let secondElement = try iterator.next()?.get() else {
+            // newElements just contains a single element
+            // and we hit the fast path
+            self = .one(firstElement)
+            return
+        }
+        
+        var elements: [Element] = []
+        elements.reserveCapacity(newElements.underestimatedCount)
+        elements.append(firstElement)
+        elements.append(secondElement)
+        while let nextElement = try iterator.next()?.get() {
+            elements.append(nextElement)
+        }
+        self = .arbitrary(elements)
+    }
+    
+    @inlinable
+    init() {
+        self = .arbitrary([])
+    }
+    
+    @inlinable
+    mutating func append(_ newElement: Element) {
+        self.append(contentsOf: CollectionOfOne(newElement))
+    }
+    
+    @inlinable
+    mutating func append(contentsOf newElements: some Sequence<Element>){
+        switch self {
+        case .one(let firstElement):
+            var iterator = newElements.makeIterator()
+            guard let secondElement = iterator.next() else {
+                // newElements is empty, nothing to do
+                return
+            }
+            var elements: [Element] = []
+            elements.reserveCapacity(1 + newElements.underestimatedCount)
+            elements.append(firstElement)
+            elements.append(secondElement)
+            elements.appendRemainingElements(from: &iterator)
+            self = .arbitrary(elements)
+            
+        case .arbitrary(var elements):
+            if elements.isEmpty {
+                if let array = newElements as? Array<Element> {
+                    // fast path
+                    self = .arbitrary(array)
```

Comment:
Have you measured this fast-path? Runtime type checks are quite expensive.

Reply:
If specialised this runtime check is eliminated and evaluated at compile time, otherwise it can be more expensive or less, depending on the size of the array.

Reply:
Alternatively we can add a specific `append(contentsOf:)` overload for `Array`  but we will not always hit it because `swift-certificates` often only operates on a generic `Sequence` type rather a concrete `Array`. 

For the optimisation I want to use it for we will at the moment never hit this fast-path because we will not have an array in the first place but a non-allocation lazy `Sequence`. Therefore I'm also open to just remove it to get this merged.

Reply:
I think for now let's just get it merged.